### PR TITLE
feat: introduce use of kapp-ctrl/packaging

### PR DESCRIPTION
**Pull request by cirocosta**
the commits tell most of the story, but here's the tl;dr on this pr:

  1. adds a `packaging` directory where we place templates and extra yaml objects
     (overlays & kapp configurations) that allow us to ship Cartographer as a
     carvel Package

  2. refactors ./hack/ci/e2e.sh and other `ci` related scripts which were not just
     meant to be run in CI to be more friendly with running locally, leaving to
     `ci` just those that assume a CI environment (like, `install-binaries`).

     - testing e2e locally now is a bit easier if you want to set up the
       envrionment once and then

  3. refactors the relase github actions workflow to be simply an automation on
     top of what we could do locally, i.e., removes any assumptions of being run
     from a CI environment.

at the end, we have a release directory (`gitignore`d) like so:

```
./release/
├── CHANGELOG.md
├── bundle.tar
└── package
    ├── package-install.yaml
    ├── package-metadata.yaml
    └── package.yaml
```

with release notes such as

<pre>
# Change Set

904ee8b Merge pull request #117 from vmware-tanzu/80-remove-comment
5eda41c Remove comment that now has a story associated
65ee7e5 Merge pull request #113 from vmware-tanzu/5-emit-results
<... manually redacted some here on github so it's not too big..>

# Installation

See https://github.com/vmware-tanzu/cartographer#installation.

# Checksums

```
bf3f222c57f3d2939a37c438f9002dbee6620ad28a415e535b50fe6b43a14c97  ./package/package-metadata.yaml
8dd3734081c38e699c73c7bb6706324b3c8cdda1f6bc1cc9c7a0a70895575b58  ./package/package-install.yaml
c04c0ae4eb9e5bff279ec7c5f837c8c3540db776c7f72dc233d1c47d0daa975d  ./package/package.yaml
0a0a4e34dd472ade52ee5e9f41c17c300ace92aef3cc8df5209871c8c45ec883  ./bundle.tar
```
</pre>

i'm not sure how useful `cartographer.yaml` there is now that we're just bundling though ... might actually get rid of it and going solely for bundle / from source for those that want to try it out using `ko` by themselves ..

---

todo:

- [x] update the installation guide to guide through the use of `bundle.tar`

### CODE REVIEW

File: hack/dist.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+set -o xtrace
```
Comment (cirocosta) [inline hack/dist.sh]: remember to get rid of this when moving to ready for review

File: packaging/package-install.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#@ load("@ytt:data", "data")
+
+---
+apiVersion: packaging.carvel.dev/v1alpha1
+kind: PackageInstall
+metadata:
+  name: #@ data.values.name + "." + data.values.domain + "." + data.values.version
+spec:
+  serviceAccountName: default
```
Comment (cirocosta) [inline packaging/package-install.yaml:26]: I think this should come from the `values.yaml`

**Comment by cirocosta** [comment]
at this point, just missing updating the commit message with the latest details

**Comment by cirocosta** [comment]
todo:

- [x] account for ytt updates
  - inventoring of ytt deps (open source licensing .. etc)
  - update `release.sh`

**Comment by cirocosta** [comment]
marking as draft once again to account for the installation based on the tarball

File: config/manager/deployment.yaml [hunk 3]
```diff
@@ -12,6 +12,19 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  name: private-registry-credentials
+  namespace: cartographer-system
+  annotations:
+    secretgen.carvel.dev/image-pull-secret: ""
+type: kubernetes.io/dockerconfigjson
+data:
+  .dockerconfigjson: e30K
```
Comment (emmjohnson) [inline config/manager/deployment.yaml:25]: Is this just a place holder?

**Review by emmjohnson — APPROVED** [review]

File: config/manager/deployment.yaml (hunk 3 above)
Comment (cirocosta) [reply config/manager/deployment.yaml:25]: yeah, evaluates to `{}` (empty auth)

Comment (cirocosta) [reply config/manager/deployment.yaml:25]: https://carvel.dev/kapp-controller/docs/latest/private-registry-auth/#package-authoring-and-placeholder-secrets

**Comment by cirocosta** [comment]
force-pushed to resolve a conflict after rebasing w/ main

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -24,32 +24,41 @@ jobs:
...
-        with:
-          fetch-depth: 0
+        with: {fetch-depth: 0}
 
-      - name: Set up Go
+      - name: set up Go
         uses: actions/setup-go@v2
-        with:
-          go-version: '~1.17.1'
+        with: {go-version: '~1.17.1'}
```
Comment (tbr11) [inline .github/workflows/release.yaml:33]: Should we switch to caret version matching? Go has committed to 1.x being compatible. `go-version: '^1.17.1'` ought to keep us on the latest minor and patch version.

File: hack/ci/install-binaries.sh [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+set -o pipefail
+
+readonly KO_VERSION=0.8.1
+readonly KO_CHECKSUM=9f004fa1c2b55ac765ec0c287ad0311a517a86299b7a633bc542f2fbbb3a4ea4
+readonly KUBERNETES_VERSION=1.19.2
+readonly KUBERNETES_CHECKSUM=fb13a93a800389029b06fcc74ab6a3b969ff74178252709a040e4756251739d2
+readonly KUTTL_VERSION=0.11.1
+readonly KUTTL_CHECKSUM=0fb13f8fbb6109803a06847a8ad3fae4fedc8cd159e2b0fd6c1a1d8737191e5f
+readonly GH_VERSION=2.0.0
+readonly GH_CHECKSUM=20c2d1b1915a0ff154df453576d9e97aab709ad4b236ce8313435b8b96d31e5c
```
Comment (tbr11) [inline hack/ci/install-binaries.sh:27]: Can these be passed in and defaulted? I foresee pain if we have to repro a build from a particular commit and a dependency version got pulled.

Comment (cirocosta) [reply hack/ci/install-binaries.sh:27]: sgtm!

File: .github/workflows/release.yaml (hunk 4 above)
Comment (cirocosta) [reply .github/workflows/release.yaml:33]: sgtm!

**Pull request merged**